KAFKA-16206 Fix unnecessary topic config deletion during ZK migration

ahuang98 opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ahuang98 committed:
wip

ahuang98 committed:
more test changes

ahuang98 committed:
remaining tests

ahuang98 committed:
comments

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

rondagostino commented on metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:
```diff
@@ -102,14 +115,18 @@ public class ClusterImageTest {
-            BrokerRegistrationInControlledShutdownChange.IN_CONTROLLED_SHUTDOWN.value()),
-            (short) 0));
+                BrokerRegistrationInControlledShutdownChange.IN_CONTROLLED_SHUTDOWN.value()),
+            FENCE_BROKER_RECORD.highestSupportedVersion()));
```
`s/FENCE_BROKER_RECORD/BROKER_REGISTRATION_CHANGE_RECORD/`

rondagostino commented on metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:
```diff
@@ -156,6 +173,66 @@ public class ClusterImageTest {
+        DELTA2_RECORDS.add(new ApiMessageAndVersion(new BrokerRegistrationChangeRecord().
+            setBrokerId(0).setBrokerEpoch(1000).setInControlledShutdown(
+                BrokerRegistrationInControlledShutdownChange.NONE.value()),
+            FENCE_BROKER_RECORD.highestSupportedVersion()));
```
`s/FENCE_BROKER_RECORD/BROKER_REGISTRATION_CHANGE_RECORD/`

rondagostino commented on metadata/src/test/java/org/apache/kafka/image/ConfigurationsImageTest.java:
```diff
@@ -61,26 +62,34 @@ public class ConfigurationsImageTest {
+        DELTA1_RECORDS.add(new ApiMessageAndVersion(new ConfigRecord().setResourceType(BROKER.id()).
+            setResourceName("1").setName("foobar").setValue(null),
+            CONFIG_RECORD.highestSupportedVersion()));
+        // add back config to b1
```
`s#// add back config to b1#// add new config to b1#`

rondagostino commented on metadata/src/test/java/org/apache/kafka/image/FeaturesImageTest.java:
```diff
@@ -96,6 +121,20 @@ public void testImage2RoundTrip() {
+    public void testApplyDelta2() {
+        assertEquals(IMAGE3, DELTA2.apply());
+        // check image2 + delta2 = image3, since records for image2 + delta2 might differ from records from image3
+        List<ApiMessageAndVersion> records = getImageRecords(IMAGE3);
```
`s/IMAGE3/IMAGE2`

I wonder why this test is still passing with this mistake?  Probably not a big deal, but still might be good to satisfy yourself that it makes sense that it would pass like this.

ahuang98 replied:
thanks for catching this, think it makes sense to me that this would pass - the records from delta2 applied ontop of image3 still come out to image3

rondagostino commented on metadata/src/test/java/org/apache/kafka/image/TopicsImageTest.java:
```diff
@@ -176,22 +200,30 @@ private PartitionRegistration newPartition(int[] replicas) {
 
-        // Create a new foo topic with a different id
-        Uuid newFooId = Uuid.fromString("b66ybsWIQoygs01vdjH07A");
+        // Attempt to create a new foo topic with a different id and replica placement
+        topicRecords.add(
+            new ApiMessageAndVersion(
+                new TopicRecord().setName("foo").setTopicId(FOO_UUID3),
+                TOPIC_RECORD.highestSupportedVersion()
+            )
+        );
+        topicRecords.add(newPartitionRecord(FOO_UUID3, 0, Arrays.asList(0, 1, localId)));
```
Could you explain this?  We already have a record creating a topic `foo` with `FOO_UUID2`.  I would expect this new record to not be generated by the controller since a topic with that name already exists.  So while I agree that `TopicsDelta` currently will replay it successfully, I'm not sure we should be testing what happens when it does since we never expect this to happen.  In fact, it could be argued that the delta should not replay it.  WDYT?

ahuang98 replied:
I had a feeling someone might comment on this 😅 I agree we should never be able to get to this point (though arguably we might now with the metadata recovery tool), my intent was to show with this test that we don't do any validation - if that changed in the future, then the test would be changed as well.
I see how it's confusing though, maybe best to remove it.

rondagostino reviewed (commented):
Thanks for the changes!  Left a few comments. Nothing major, though the TopicsImageTest question is an interesting one, I think.

ahuang98 commented on metadata/src/test/java/org/apache/kafka/image/ProducerIdsImageTest.java:
```diff
@@ -54,11 +54,24 @@ public class ProducerIdsImageTest {
         RecordTestUtils.replayAll(DELTA1, DELTA1_RECORDS);
 
-        IMAGE2 = new ProducerIdsImage(789);
+        // producer ids must be increasing, but delta replay does not check this (only QuorumController replay)
```
@rondagostino what are your thoughts on this check then? should I remove it as well?

rondagostino replied:
I guess we need to decide what to do with these types of cases.  The overall issue is that right now the various `*Delta` classes will replay pretty much whatever you give them without complaining -- including stuff that the actual `QuorumController` would never send and that would be unable to replay.

We could keep these changes as opposed to just silently deleting them from this PR/ignoring the issue that they point to.  If we did that then at a minimum we should add copious comments about what the test is doing and why we expect it to never happen/why we are testing it anyway.

But it feels to me that these are strange cases to test, and it might just be best to open a JIRA ticket to make these various `*Delta` classes actually reject illegal stuff like this so that we get an exception closer to the point where the issue actually originates.

ahuang98 replied:
let's go with the latter then, we'll need to iron out validation for the sake of the metadata recovery tool anyways

ahuang98 committed:
addressing ron's comments

rondagostino reviewed (approved):
Looks good -- thanks for the changes!

Can you merge in latest changes from trunk to get a new build?  There were lots of test failures with the last build, and it seems safest to observe those go away if possible.

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

rondagostino commented:
@ahuang98 I still saw lots of test failures, and when I checked them locally I actually was able to reproduce the below failures.  Can you take a look to see if you also agree these fail locally and then, assuming you agree, investigate as to why?
```
testTopicDualWriteDelta()
testTopicDualWriteSnapshot()
testControllerFailover()
```

ahuang98 committed:
test fixes

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 commented on metadata/src/test/java/org/apache/kafka/metadata/migration/KRaftMigrationDriverTest.java:
```diff
@@ -627,10 +629,13 @@ public void testTopicDualWriteDelta() throws Exception {
+            assertTrue(topicClient.createdTopics.contains("baz"));
             assertTrue(topicClient.updatedTopicPartitions.get("bar").contains(0));
-            assertEquals(new ConfigResource(ConfigResource.Type.TOPIC, "foo"), configClient.deletedResources.get(0));
+            assertEquals(2, configClient.deletedResources.size());
```
It's strange that `configClient` tracks two deleted resources while `topicClient` only sees one. 
The only difference between this and the former test seems to be that the migration driver calls `onMetadataUpdate` on a `LogDeltaManifest` here rather than a `SnapshotManifest` in the former test.
cc @mumrah not sure if there's a bug here or if I'm missing something

ahuang98 committed:
fix

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
test changes

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
retrigger build

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 commented:
Latest test run has 14 "new failing" which are all timeout related/unrelated - https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14206/17/tests

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

mumrah reviewed (commented):
Just writing this all out to make sure I'm understanding the various scenarios.

## Topic with no configs

### Delta case

A topic is deleted which has no configs. The new logic prevents ConfigurationDelta from being created. This means KRaftMigrationZkWriter#handleDelta only processes the TopicsDelta and calls ZkTopicMigrationClient#deleteTopic which deletes the topic znode and config znode. Note that the config znode should always exist for a topic even if there are no configs (@mimaison fixed this recently #15022). This case is fixed by this PR ✅ 

### Snapshot case

Same as above, but we call KRaftMigrationZkWriter#handleTopicsSnapshot. For the deleted topic, we call ZkTopicMigrationClient#deleteTopic. Prior to this patch we were also calling ZkConfigMigrationClient#deleteConfigs which was logging an error because of NO_NODE (since the topic znode was already deleted at this point). This case is fixed by this PR ✅ 

## Topic with configs

### Delta case

A topic is deleted which has configs. A TopicDelta and ConfigurationDelta are created. KRaftMigrationZkWriter calls handleTopicsDelta before handleConfigsDelta. handleConfigsDelta will see a NO_NODE error when trying to delete the configs. This is one of the sources of the error we are seeing in KAFKA-16206. 

### Snapshot case

Since ConfigurationsImage doesn't track deleted resources like TopicsImage does, I think this case is fine. ✅ 


------------------

If I'm understanding things correctly, we still need to handle the delta case for a topic being deleted that has configs. It's a bit tricky actually since we can't just reorder things in KRaftMigrationZkWriter#handleDelta. If we moved configuration first, we might run into problems with the topic config znode not existing. 

Maybe the simplest thing here is to change the NO_NODE log message from `error` to `debug` in ZkConfigMigrationClient#deleteConfigs. 

@mimaison any opinions here (since you looked a this code recently)?

ahuang98 commented:
Is there a downside to having `deleteTopic` in `ZkTopicMigrationClient` not delete configs? Otherwise changing the logging level seems okay to me.

ahuang98 committed:
updated logging

ahuang98 commented:
@mimaison tagging you again just in case you have time to take a look
@mumrah I've added the logging change in the meantime - are we okay merging in the next week if we don't get a response?

mimaison commented on core/src/main/scala/kafka/zk/migration/ZkConfigMigrationClient.scala:
```diff
@@ -226,8 +226,8 @@ class ZkConfigMigrationClient(
       if (responses.head.resultCode.equals(Code.NONODE)) {
-        // Not fatal.
-        error(s"Did not delete $configResource since the node did not exist.")
+        // Not fatal. This is expected in the case this is a topic config and we delete the topic (KAFKA-16206)
```
We usually don't include mention to Jiras. The explanation seems sufficient.

mimaison commented on metadata/src/test/java/org/apache/kafka/image/ClientQuotasImageTest.java:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.kafka.image;
 
+import java.util.Collections;
```
Can we move this with the other `java.util` imports?

mimaison commented on metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:
```diff
@@ -17,6 +17,11 @@
 
 package org.apache.kafka.image;
 
+import org.apache.kafka.common.metadata.RegisterBrokerRecord;
```
Can we move these imports with the other `org.apache.kafka.common.metadata` imports below?

mimaison commented on metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:
```diff
@@ -64,6 +71,12 @@ public class ClusterImageTest {
+
+    static final ClusterDelta DELTA2;
+
+    final static ClusterImage IMAGE3;
```
nit: `static final` is the preferred syntax

mimaison commented on metadata/src/test/java/org/apache/kafka/image/ConfigurationsImageTest.java:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.kafka.image;
 
+import java.util.Collections;
```
Let's move this below

mimaison reviewed (commented):
Thanks for the PR. I left a few suggestions. 

A lot of the test changes don't seem directly related to the fix. It would be better to move them in a separate PR.

ahuang98 commented:
@mimaison the PR was originally meant to introduce additional test changes, I believe @mumrah renamed it after I added the migration fix. I'll move the tests out and apply your suggestions, thanks!

ahuang98 committed:
reverting test changes unrelated to kafka-16206

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
comments

ahuang98 commented:
@mimaison @mumrah I moved unrelated test changes over to https://github.com/apache/kafka/pull/15373. The latest commit [1c1eb7d](https://github.com/apache/kafka/pull/15373/commits/1c1eb7d1a866c56188a97f6fa41940933e456c03) addresses the comments left here (mostly imports)

mimaison commented:
There's quite a few failures related to ZkMigration in the last CI run: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14206/22/testReport/

ahuang98 committed:
Merge branch 'trunk' into image_test_improvements

ahuang98 committed:
topicsImageTest

ahuang98 commented:
Thanks @mimaison, looks like this is because I split out the TopicsImageTest changes into a separate PR. The migration tests depend on the `DELTA1_RECORDS` defined there so I moved over changes from TopicsImageTest as well.

ahuang98 committed:
retrigger build

ahuang98 commented:
@mimaison could you help merge this if this looks good to you? :)
